ArcGISRuntimeSDKQt_CppSamples/Layers/ListKmlContents/ListKmlContents.cpp

```diff
@@ -210,6 +215,95 @@ void ListKmlContents::processSelectedNode(const QString& nodeName)
   }
 }
 
+void ListKmlContents::getViewpointFromKmlViewpoint(KmlNode* node)
+{
+  const KmlViewpoint kmlViewpoint = node->viewpoint();
+
+  if (!kmlViewpoint.isEmpty())
+  {
+    // altitude adjustment is needed for all but Absolute altitude mode
+    m_needsAltitudeFixed = (kmlViewpoint.altitudeMode() != KmlAltitudeMode::Absolute);
+
+    switch (kmlViewpoint.type())
+    {
+    case KmlViewpointType::LookAt:
+      m_viewpoint = Viewpoint(kmlViewpoint.location(), Camera(kmlViewpoint.location(), kmlViewpoint.range(),
+                                                              kmlViewpoint.heading(), kmlViewpoint.pitch(), kmlViewpoint.roll()));
+      return;
+    case KmlViewpointType::Camera:
+      m_viewpoint = Viewpoint(kmlViewpoint.location(), Camera(kmlViewpoint.location(),
+                                                              kmlViewpoint.heading(), kmlViewpoint.pitch(), kmlViewpoint.roll()));
+      return;
+    default:
+      qWarning("Unexpected KmlViewpointType");
+      return;
+    }
+  }
+
+  // if viewpoint was empty, then use node's extent
+  const Envelope nodeExtent = node->extent();
+  if (nodeExtent.isValid() && !nodeExtent.isEmpty())
+  {
+    // when no altitude is specified, assume elevation needs to be adjusted
+    m_needsAltitudeFixed = true;
+
+    if (nodeExtent.width() == 0 && nodeExtent.height() == 0)
+    {
+      // default values based on Google Earth
+      m_viewpoint = Viewpoint(nodeExtent);
+      m_sceneView->setViewpointCameraAndWait(Camera(nodeExtent.center(), 1000, 0, 45, 0));
+      return;
+    }
+    else
+    {
+      // add padding to extent
+      double bufferDistance = qMax(nodeExtent.width(), nodeExtent.height()) / 20;
+      Envelope bufferedExtent = Envelope(nodeExtent.xMin() - bufferDistance, nodeExtent.yMin() - bufferDistance,
+                                         nodeExtent.xMax() + bufferDistance, nodeExtent.yMax() + bufferDistance,
+                                         nodeExtent.zMin() - bufferDistance, nodeExtent.zMax() + bufferDistance,
+                                         SpatialReference::wgs84());
+      m_viewpoint = Viewpoint(bufferedExtent);
+    }
+  }
+  else
+  {
+    // can't show viewpoint
+    m_needsAltitudeFixed = false;
+    m_viewpoint = Viewpoint();
+  }
+}
+
+void ListKmlContents::getAltitudeAdjustedViewpoint(KmlNode* node)
+{
+  // assume altitude mode is clamp-to-ground if not specified
+  KmlAltitudeMode altMode = KmlAltitudeMode::ClampToGround;
+  KmlViewpoint kmlViewpoint = node->viewpoint();
+
+  if (!kmlViewpoint.isEmpty())
+  {
+    altMode = kmlViewpoint.altitudeMode();
+  }
+
+  // if altitude mode is Absolute, viewpoint doesn't need adjustment
+  if (altMode == KmlAltitudeMode::Absolute)
+  {
+    m_sceneView->setViewpoint(m_viewpoint);
+    return;
+  }
+
+  const Envelope lookAtExtent = static_cast<Envelope>(m_viewpoint.targetGeometry());
```

Comment:
you should probably use the `geometry_cast` helper here instead. And I think that we can't use the `targetGeometry` for both. Maybe the envelope should be `targetGeometry().extent()`?

Reply:
I think it should be as is because otherwise the cast to `Envelope` seems to always be valid, even in cases where it should be cast to a `Point`